pxr/imaging/plugin/hdEmbree/ies.cpp

```diff
@@ -0,0 +1,490 @@
+    clear();
+
+    if (!parse(ies) || !process()) {
+        clear();
+        return false;
+    }
+
+    return true;
+}
+
+void IESFile::clear() {
+    _intensities.clear();
+    _v_angles.clear();
+    _h_angles.clear();
+}
+
+int IESFile::packed_size() {
+    if (_v_angles.size() && _h_angles.size() > 0) {
+        return 2 + _h_angles.size() + _v_angles.size() +
+                _h_angles.size() * _v_angles.size();
+    }
+    return 0;
+}
+
+template <class To, class From>
+std::enable_if_t<sizeof(To) == sizeof(From) &&
+                    std::is_trivially_copyable_v<From> &&
+                    std::is_trivially_copyable_v<To>,
+                    To>
+    // constexpr support needs compiler magic
+    bit_cast(const From &src) noexcept {
+    static_assert(std::is_trivially_constructible_v<To>,
+                "This implementation additionally requires "
+                "destination type to be trivially constructible");
+
+    To dst;
+    std::memcpy(&dst, &src, sizeof(To));
+    return dst;
+}
+
+void IESFile::pack(float *data) {
+    if (_v_angles.size() && _h_angles.size()) {
+        *(data++) = bit_cast<int, float>(_h_angles.size());
+        *(data++) = bit_cast<int, float>(_v_angles.size());
+
+        memcpy(data, &_h_angles[0], _h_angles.size() * sizeof(float));
+        data += _h_angles.size();
+        memcpy(data, &_v_angles[0], _v_angles.size() * sizeof(float));
+        data += _v_angles.size();
+
+        for (int h = 0; h < _intensities.size(); h++) {
+            memcpy(data, &_intensities[h][0], _v_angles.size() * sizeof(float));
+            data += _v_angles.size();
+        }
+    }
+}
+
+class IESTextParser {
+public:
+    std::vector<char> text;
+    char *data;
+    bool error;
+
+    IESTextParser(const std::string &str)
+        : text(str.begin(), str.end()), error(false) {
+        std::replace(text.begin(), text.end(), ',', ' ');
+        data = strstr(&text[0], "\nTILT=");
+    }
+
+    bool eof() { return (data == NULL) || (data[0] == '\0'); }
+
+    bool has_error() { return error; }
+
+    double get_double() {
+        if (eof()) {
+            error = true;
+            return 0.0;
+        }
+        char *old_data = data;
+        double val = strtod(data, &data);
+        if (data == old_data) {
+            data = NULL;
+            error = true;
+            return 0.0;
+        }
+        return val;
+    }
+
+    long get_long() {
+        if (eof()) {
+            error = true;
+            return 0;
+        }
+        char *old_data = data;
+        long val = strtol(data, &data, 10);
+        if (data == old_data) {
+            data = NULL;
+            error = true;
+            return 0;
+        }
+        return val;
+    }
+};
+
+bool IESFile::parse(const std::string &ies) {
+    if (ies.empty()) {
+        return false;
+    }
+
+    IESTextParser parser(ies);
+    if (parser.eof()) {
+        return false;
+    }
+
+    // Handle the tilt data block
```

Comment:
have you scanned past the IESNA:XXXX line somewhere? It's common for the first line of an IES file to be

```
IESNA:LM-63-2002
IESNA:LM-63-1995
IESNA91
IESNA:LM-63-1991
IESNA:LM-63-1986
```
If the line isn't there, then the file is by default assumed to be IESNA:LM-63-1986. My IES test corpus commonly had all of the above, but there isn't any impact on the parser here. Mostly, if the `IESNA*` string appears in the first line it should be skipped.

Reply:
This is handled by the constructor of `IESTextParser`, it skips directly to the first occurrence of `\nTILT=`.